module/zfs/dmu.c

```diff
@@ -964,6 +967,62 @@ dmu_free_range(objset_t *os, uint64_t object, uint64_t offset,
 	return (0);
 }
 
+static void
+dmu_read_abd_done(zio_t *zio)
+{
+    abd_put((abd_t *)zio->io_private);
+}
+
+static int
+dmu_read_abd(dnode_t *dn, uint64_t offset, uint64_t size,
+    abd_t *data, uint32_t flags)
+{
+    spa_t *spa = dn->dn_objset->os_spa;
+    dmu_buf_t **dbp;
+    int numbufs, err, i;
+    size_t off = 0;
+    zio_t *rio;
+
+    /*
+     * Direct IO must be block aligned
+     */
+     ASSERT(flags & DMU_DIRECTIO);
+     ASSERT(offset % dn->dn_datablksz == 0);
+     ASSERT(size % dn->dn_datablksz == 0);
```

Comment:
I think, from the read side, this is pretty straightforward. We simply need to "pad out" the abd buffer with allocated buffer space to align with the block size. For the write side, it's a bit trickier... but still possible I think: we can issue reads (ala read-modify-write) via dmu_will_fill() for alignment problems at the beginning and end of the range, and then construct an abd buffer by pulling the missing write data from these. Both of these may require the abd MULTI_LIST feature I added (to construct a meta-abd buffer from multiple abd buffers) if we want to avoid extra copies.